pnp: move shadow loggers to parent module

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92058
* #91639
* __->__ #91428
* #91421

Summary:

Before this PR, PNP added shadow loggers to insides of
the shadow wrapper modules.

This PR moves those loggers to the parent module.

There are a couple of benefits:
1. this will unbreak features of quantization API which don't support loggers (such as hardcoding model output to be quantized)
2. this makes it easier to look at the parent graph and visualize what is logged, since now all the logging is in the same graph
3. this will make it easier to implement features such as propagation error calculation in the future

Test plan:

```
python test/test_quantization.py -k NShadows
```

vkuzo committed:
pnp: move shadow loggers to parent module

Summary:

Before this PR, PNP added shadow loggers to insides of
the shadow wrapper modules.

This PR moves those loggers to the parent module.

There are a couple of benefits:
1. this will unbreak features of quantization API which don't support loggers (such as hardcoding model output to be quantized)
2. this makes it easier to look at the parent graph and visualize what is logged, since now all the logging is in the same graph
3. this will make it easier to implement features such as propagation error calculation in the future

Test plan:

```
python test/test_quantization.py -k NShadows
```

[ghstack-poisoned]

vkuzo committed:
Update on "pnp: move shadow loggers to parent module"


Summary:

Before this PR, PNP added shadow loggers to insides of
the shadow wrapper modules.

This PR moves those loggers to the parent module.

There are a couple of benefits:
1. this will unbreak features of quantization API which don't support loggers (such as hardcoding model output to be quantized)
2. this makes it easier to look at the parent graph and visualize what is logged, since now all the logging is in the same graph
3. this will make it easier to implement features such as propagation error calculation in the future

Test plan:

```
python test/test_quantization.py -k NShadows
```

[ghstack-poisoned]

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -593,16 +523,14 @@ def handle_subgraph_candidate(
-            # node of this subgraph
-            new_args = [last_node]
-
+            new_args = []
```
just for my understanding, why is the `last_node` argument removed?

vkuzo replied:
Before it was

```
# parent
x0 -> op -> x1 
  \                    \
    -----------shadow

# shadow
x0 -> op_shadow -> x1_shadow -> log 
                                                            /
x1 --------------------------------
```

Now, it's

```
# parent
x0 -> op -> x1 ---------> log
  \                                         /  
    -shadow -> x1_shadow


# shadow
x0 -> op_shadow -> x1_shadow
```

the logger is moved outside, and `last_node` is the thing which the logger needs to compare actual to reference

jerryzh168 reviewed (approved):
looks good, had some inline comments

vkuzo commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -705,11 +647,21 @@ def handle_subgraph(
 
     fqn = _maybe_get_fqn(first_node, mt)
 
+    # We want the results to contain the subgraphs in natural order,
```
just tried it, unfortunately it leads to the same issue explained below in the comment :(

vkuzo committed:
Update on "pnp: move shadow loggers to parent module"


Summary:

Before this PR, PNP added shadow loggers to insides of
the shadow wrapper modules.

This PR moves those loggers to the parent module.

There are a couple of benefits:
1. this will unbreak features of quantization API which don't support loggers (such as hardcoding model output to be quantized)
2. this makes it easier to look at the parent graph and visualize what is logged, since now all the logging is in the same graph
3. this will make it easier to implement features such as propagation error calculation in the future

Test plan:

```
python test/test_quantization.py -k NShadows
```

[ghstack-poisoned]

vkuzo committed:
Update on "pnp: move shadow loggers to parent module"


Summary:

Before this PR, PNP added shadow loggers to insides of
the shadow wrapper modules.

This PR moves those loggers to the parent module.

There are a couple of benefits:
1. this will unbreak features of quantization API which don't support loggers (such as hardcoding model output to be quantized)
2. this makes it easier to look at the parent graph and visualize what is logged, since now all the logging is in the same graph
3. this will make it easier to implement features such as propagation error calculation in the future

Test plan:

```
python test/test_quantization.py -k NShadows
```

[ghstack-poisoned]

vkuzo commented:
@pytorchbot merge

vkuzo committed:
Update on "pnp: move shadow loggers to parent module"


Summary:

Before this PR, PNP added shadow loggers to insides of
the shadow wrapper modules.

This PR moves those loggers to the parent module.

There are a couple of benefits:
1. this will unbreak features of quantization API which don't support loggers (such as hardcoding model output to be quantized)
2. this makes it easier to look at the parent graph and visualize what is logged, since now all the logging is in the same graph
3. this will make it easier to implement features such as propagation error calculation in the future

Test plan:

```
python test/test_quantization.py -k NShadows
```

[ghstack-poisoned]

vkuzo commented:
@pytorchbot merge --help

vkuzo commented:
@pytorchbot --help

vkuzo commented:
@pytorchbot merge

vkuzo commented:
@pytorchmergebot --help

vkuzo commented:
@pytorchbot merge -f "failures are unrelated to PR"
